### PR TITLE
feature : Alerter for Dialog

### PR DESCRIPTION
**Pull request by ShwetaChauhan18**
Solution for this issue : https://github.com/Tapadoo/Alerter/issues/247

I need this feature in urgent base, So if it's good if you review and check this feature. Otherwise I have to add module in my project. 

Also, I can refactor code in next PR.

### CODE REVIEW

**Comment by kpmmmurphy** [comment]
Hey @ShwetaChauhan18, thanks for your PR, and great to see an implementation for this! There is a lot of code duplication in the AlertForDialog class, which appears to be directly copied from the original Alert class. I think an inheritance approach would be better, as it would reduce this duplication. I'm not happy to merge this in at the moment, so I do recommend building your fork locally and including it as a module. I'll provide a few comments on the changed files and we can take it from there! Thanks again.

File: alerter/src/main/java/com/tapadoo/alerter/AlerterForDialog.kt [hunk 1]
```diff
@@ -0,0 +1,817 @@
...
+        private var mAppCompatDialog: AppCompatDialog? = null
+
+
+        @JvmStatic
+        fun create(context: Context?, view: AppCompatDialog): AlerterForDialog {
+            return create(context, R.layout.alerter_alert_default_layout, view)
+        }
+
+        @JvmStatic
+        fun create(context: Context?, @LayoutRes layoutId: Int, view: AppCompatDialog): AlerterForDialog {
```
Comment (kpmmmurphy) [inline alerter/src/main/java/com/tapadoo/alerter/AlerterForDialog.kt:754]: I think the right approach would be to refactor this into the original Alerter class, allowing the caller to use either and activity, or dialog to display the alert.

**Review by kpmmmurphy — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/tapadoo/alerter/demo/CustomBottomSheetDialogFragment.kt [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+import kotlinx.android.synthetic.main.layout_modal_bottom_sheet.secondButton
+import kotlinx.android.synthetic.main.layout_modal_bottom_sheet.thirdButton
+
+class CustomBottomSheetDialogFragment : BottomSheetDialogFragment() {
+
+    companion object {
+        const val TAG = "CustomBottomSheetDialogFragment"
+    }
+
+    var mAppCompatDialog: AppCompatDialog? = null
```
[the remark this answers is not in the document]
Comment (ShwetaChauhan18) [reply app/src/main/java/com/tapadoo/alerter/demo/CustomBottomSheetDialogFragment.kt]: done

File: app/src/main/AndroidManifest.xml [hunk 3]
```diff
@@ -23,6 +23,11 @@
             android:name=".KotlinDemoActivity"
             android:label="@string/title_activity_example"
             android:theme="@style/AppTheme">
+        </activity>
+
+        <activity android:name=".MainActivity"
```
[the remark this answers is not in the document]
Comment (ShwetaChauhan18) [reply app/src/main/AndroidManifest.xml]: changed

File: app/src/main/java/com/tapadoo/alerter/demo/MainActivity.kt [hunk 4]
```diff
@@ -0,0 +1,20 @@
+package com.tapadoo.alerter.demo
+
+import android.os.Bundle
+import androidx.appcompat.app.AppCompatActivity
+import kotlinx.android.synthetic.main.activity_main.btnBottomSheetModal
+
+class MainActivity : AppCompatActivity() {
```
[the remark this answers is not in the document]
Comment (ShwetaChauhan18) [reply app/src/main/java/com/tapadoo/alerter/demo/MainActivity.kt]: changed

File: alerter/src/main/java/com/tapadoo/alerter/AlerterForDialog.kt (hunk 1 above)
Comment (ShwetaChauhan18) [reply alerter/src/main/java/com/tapadoo/alerter/AlerterForDialog.kt]: changed

**Comment by ShwetaChauhan18** [comment]
> Hey @ShwetaChauhan18, thanks for your PR, and great to see an implementation for this! There is a lot of code duplication in the AlerterForDialog class, which appears to be directly copied from the original Alerter class. I think an inheritance approach would be better, as it would reduce this duplication. I'm not happy to merge this in at the moment, so I do recommend building your fork locally and including it as a module. I'll provide a few comments on the changed files and we can take it from there! Thanks again.

changed

File: alerter/src/main/java/com/tapadoo/alerter/Alerter.kt [hunk 5]
```diff
@@ -774,7 +793,13 @@ class Alerter private constructor() {
          */
         @JvmStatic
         fun clearCurrent(activity: Activity?) {
-            (activity?.window?.decorView as? ViewGroup)?.let {
```
Comment (kpmmmurphy) [inline alerter/src/main/java/com/tapadoo/alerter/Alerter.kt:777]: Please remove Hungarian notation.

File: alerter/src/main/java/com/tapadoo/alerter/Alerter.kt [hunk 6]
```diff
@@ -745,6 +752,17 @@ class Alerter private constructor() {
             return create(activity, R.layout.alerter_alert_default_layout)
         }
 
+        /**
+         * Creates the Alert, and maintains a reference to the calling Activity
+         *
+         * @param activity The calling Activity
```
Comment (kpmmmurphy) [inline alerter/src/main/java/com/tapadoo/alerter/Alerter.kt:767]: Please update the Kotlin docs.

**Comment by ShwetaChauhan18** [comment]
@kpmmmurphy : Please Review. I have refactored all code.

File: alerter/src/main/java/com/tapadoo/alerter/Alerter.kt [hunk 7]
```diff
@@ -753,15 +780,18 @@ class Alerter private constructor() {
          * @return This Alerter
          */
         @JvmStatic
-        fun create(activity: Activity?, @LayoutRes layoutId: Int): Alerter {
+        fun create(activity: Activity?, @LayoutRes layoutId: Int, dialog: AppCompatDialog? = null): Alerter {
```
Comment (kpmmmurphy) [inline alerter/src/main/java/com/tapadoo/alerter/Alerter.kt:783]: I think this makes for a confusing API. When we have a dialog, there is no need for the Activity. Ultimately, all we require is a reference to the DecorView, in order to add, and remove the Alert. Perhaps storing a weak ref to the only the DecorView is the optimal solution here

**Review by kpmmmurphy — CHANGES_REQUESTED** [review]

**Comment by kpmmmurphy** [comment]
Thanks for updating your PR @ShwetaChauhan18, we're definitely moving in the right direction. I think if you refactor the Alerter to only maintain a reference to the DecorView, it will be ready to merge! Don't forget to update the version in the build.gradle, and the changelog! I think it's best to move to version 7.0.0. Thanks again.

File: alerter/src/main/java/com/tapadoo/alerter/Alerter.kt (hunk 6 above)
Comment (ShwetaChauhan18) [reply alerter/src/main/java/com/tapadoo/alerter/Alerter.kt:767]: Changed

**Comment by ShwetaChauhan18** [comment]
PR updated as per comment

**Comment by ShwetaChauhan18** [comment]
@kpmmmurphy : Pr Updated as per comment. Please Review it and Thanks a lot for quick response.

**Comment by kpmmmurphy** [comment]
Hey @ShwetaChauhan18, I've created a new PR (#251) based of your code here. I refactored to improve the API, creating a cleaner calling surface. Thanks again for your help. I'm going to close this one out. One important thing to note, is that force pushing clears the PRs history, making it hard for reviewers to spot code changes for different commits - in future I'd recommend trying to maintain the commit history, as it makes reviewing easier! 🙌

**Pull request closed without merging**